src/systems/logical_audio_sensor_plugin/Source.cc

```diff
@@ -0,0 +1,255 @@
+/*
+ * Copyright (C) 2020 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "Source.hh"
+
+#include <algorithm>
+#include <cctype>
+#include <string>
+#include <unordered_map>
+
+namespace ignition
+{
+namespace gazebo
+{
+// Inline bracket to help doxygen filtering.
+inline namespace IGNITION_GAZEBO_VERSION_NAMESPACE {
+namespace systems
+{
+namespace logical_audio
+{
+  /// \brief A map to help convert user-input strings to the proper
+  ///   attenuation function.
+  const std::unordered_map<std::string, AttenuationFunction>
+    kAttFuncMap {{"linear", AttenuationFunction::Linear}};
+
+  /// \brief A map to help convert user-input strings to the proper
+  ///   attenuation shape.
+  const std::unordered_map<std::string, AttenuationShape>
+    kAttShapeMap {{"sphere", AttenuationShape::Sphere}};
+
+  //////////////////////////////////////////////////
+  Source::Source(const unsigned int _id,
+                 const ignition::math::Vector3d &_position,
+                 const AttenuationFunction _attenuationFunc,
+                 const AttenuationShape _attenuationShape,
+                 const float _innerRadius,
+                 const float _falloffDistance,
+                 const float _volumeLevel,
+                 const bool _playing) :
+    id(_id),
+    position(_position),
+    attenuationFunc(_attenuationFunc),
+    attenuationShape(_attenuationShape),
+    playing(_playing)
+  {
+    this->SetInnerRadius(_innerRadius);
+    this->SetFalloffDistance(_falloffDistance);
+    this->SetVolumeLevel(_volumeLevel);
+  }
+
+  //////////////////////////////////////////////////
+  Source::Source(const unsigned int _id,
```

Comment:
This looks like a good place for a [delegating constructor](https://www.geeksforgeeks.org/constructor-delegation-c/)